design/design-doc-external-secret-stores.md

```diff
@@ -0,0 +1,655 @@
+interface, we will receive this configuration as arbitrary key value pairs with
+an optional `attributes` field.
+
+**Examples:**
+
+Publish Connection details to a _Kubernetes_ secret (already existing case):
+
+```yaml
+spec:
+  publishConnectionDetailsTo:
+    name: my-db-connection
+    kubernetes:
+      namespace: crossplane-system
+```
+
+Publish Connection details to a _Vault_ secret:
+
+```yaml
+spec:
+  publishConnectionDetailsTo:
+    name: my-db-connection
+    externalStore:
+      configRef:
+        name: vault-default
+        namespace: crossplane-system
+```
+
+Publish Connection details to a _Kubernetes_ secret with some labels:
+
+```yaml
+spec:
+  publishConnectionDetailsTo:
+    name: my-db-connection
+    kubernetes:
+      namespace: crossplane-system
+      labels:
+        environment: production
+      annotations:
+        acme.example.io/secret-type: infrastructure
+```
+
+Publish Connection details to an _AWS Secret Manager_ secret with some tags:
+
+```yaml
+spec:
+  publishConnectionDetailsTo:
+    name: my-db-connection
+    externalStore:
+      configRef:
+        name: aws-secret-manager-platform
+        namespace: crossplane-system
+      attributes:
+        tags:
+          environment: production
+```
+
+Publish Connection details to _Vault_ but **still store it as a Kubernetes
+secret**:
+
+```yaml
+spec:
+  publishConnectionDetailsTo:
+    name: my-db-connection
+    kubernetes:
+      namespace: crossplane-system
+    externalStore:
+      configRef:
+        name: vault-platform
+        namespace: crossplane-system
+```
+
+#### Separation of Concerns: Compositions, Composites and Claims
+
+`StoreConfig` resource will be a _namespaced_ resource to allow platform
+consumers to configure their own secret store of choice. This is aligned with
+the existing approach since Kubernetes secrets are namespaced resources and
+platform operators could specify where should composite resource secrets land
+using the partial input `writeConnectionSecretsToNamespace`. We will similarly
+deprecate this field in favor of `publishConnectionDetailsToStore`.
+
+**Examples:**
+
+**Composition** configuring the Kubernetes namespace that connection secrets
+for Composite resources will be stored:
+
+```yaml
+spec:
+  publishConnectionDetailsToStore:
+    kubernetes:
+      namespace: crossplane-system
+```
+
+**Composition** configuring the secret store that connection secrets for
+Composite resources will be stored:
+
+```yaml
+spec:
+  publishConnectionDetailsToStoreConfigRef:
+    name: vault-platform
+    namespace: crossplane-system
+```
+
+**Claim** spec for writing connection secret in the same Kubernetes namespace:
+
+```yaml
+spec:
+  publishConnectionDetailsTo:
+    name: database-creds 
+    # when neither `kubernetes` nor `externalStore` configs
+    # provided, it defaults to "kubernetes in the same
+    # namespace", same as "kubernetes: {}"
```

Comment:
We will lose the ability to write both to a Kubernetes secret and to an external one: https://github.com/crossplane/crossplane/pull/2494#discussion_r733602782

Other than that this will definitely be a simpler API.

Reply:
Done.